File: mobile/lib/domain/liability_guard.dart
/// Garde-fou déterministe anti-attribution de responsabilité - miroir Dart
/// de `backend/app/domain/liability_guard.py`.
///
/// Depuis R0.1 (pivot Local-First), le Reserve Composer tourne sur
/// l'appareil (voir `lib/domain/reserve_composer.dart`) pour permettre la
/// génération du dossier hors ligne (point 6 de la demande corrective). Ce
/// garde-fou doit donc exister ICI, côté Dart, et pas seulement côté
/// backend : sans lui, le bug corrigé en R0.1 (packagingCondition =
/// "transporteur responsable" traversant jusqu'à la réserve finale) serait
/// réintroduit dès qu'un utilisateur génère une réserve sans connexion.
///
/// Les deux implémentations (Python et Dart) DOIVENT rester alignées :
/// mêmes catégories de violation, mêmes champs surveillés. Toute
/// modification ici doit être répercutée dans
/// `backend/app/domain/liability_guard.py` et vice-versa (voir
/// docs/architecture.md, section "duplication contrôlée du Reserve
/// Composer").
///
/// Déterministe : expressions régulières figées, aucun appel réseau/IA.
/// Volontairement PAS de sanitization automatique : on rejette et on
/// demande à l'utilisateur de reformuler en constat factuel neutre.
library;

import 'errors/domain_errors.dart';
import 'fact_set/confirmed_fact_data.dart';

/// Champs texte libre de [ConfirmedFactData] susceptibles de contenir une
/// formulation utilisateur non contrainte.
const List<String> liabilityGuardFreeTextFields = <String>[
  'packagingCondition',
  'productCondition',
  'locationOnItem',
  'productLabel',
  'productReference',
];

class ForbiddenPattern {
  const ForbiddenPattern(this.violationCode, this.pattern);
  final String violationCode;
  final RegExp pattern;
}

// Motifs volontairement larges (faux positifs > faux négatifs) : ces champs
// ne doivent décrire QU'UN état physique constaté ("carton enfoncé"), jamais
// une cause, une conséquence juridique/financière, ou une partie
// responsable.
//
// R0.2 (hotfix decouvert par execution reelle du test suite sur un poste
// equipe - CORRIGE UNE SECONDE FOIS apres un premier correctif insuffisant,
// voir CHANGELOG.md) : `\b` de `package:dart:core` RegExp ne reconnait QUE
// les caracteres ASCII comme "mots" - CE COMPORTEMENT N'EST PAS CHANGE par
// `unicode: true` (contrairement a ce qu'un premier correctif supposait a
// tort : ce flag active seulement `\p{...}`/`\P{...}`, il ne rend PAS `\b`
// ou `\w` sensibles a l'Unicode - meme comportement qu'en JavaScript, dont
// Dart hérite la semantique RegExp). Consequence reelle, prouvee par des
// tests qui echouaient encore APRES le premier correctif ("unicode: true"
// seul) : "carton à la charge du fournisseur" (LIABILITY_ATTRIBUTION, le
// motif commence par "à"), "remboursement dû au client"
// (INDEMNIFICATION_PROMISE, le motif "dû" se termine par "û"), "vice caché
// constaté" (LEGAL_CONCLUSION, le motif "caché" se termine par "é") -
// autrement dit le garde-fou laissait passer silencieusement exactement le
// type de formulation qu'il est censé bloquer, des que la formulation
// commençait ou finissait par une lettre accentuee.
//
// Correctif REEL : remplacer `\b` par des lookaround explicites sur la
// categorie Unicode "Lettre" (`\p{L}`) + chiffre/underscore, qui EUX sont
// bien Unicode-aware des lors que `unicode: true` est actif (c'est là son
// vrai role). `(?<![\p{L}\p{N}_])` = pas précédé d'un caractère de mot
// Unicode ; `(?![\p{L}\p{N}_])` = pas suivi d'un caractère de mot Unicode.
// R2 : rendu public (était `_forbiddenPatterns`) pour être réutilisé par
// `lib/domain/candidate_guard.dart` - même principe que le backend, où
// `app/domain/liability_guard.py::_FORBIDDEN_PATTERNS` a été renommé
// `FORBIDDEN_PATTERNS` pour la même raison (source unique de vérité sur "ce
// qu'est un contenu interdit", partagée entre le filtrage post-confirmation
// et le filtrage pré-revue des candidats IA).
final List<ForbiddenPattern> forbiddenPatterns = <ForbiddenPattern>[
  ForbiddenPattern(
    'LIABILITY_ATTRIBUTION',
    RegExp(
      r'(?<![\p{L}\p{N}_])(responsab(le|ilité)s?|fautifs?|faute|'
      r'imputable(\s+(au|à|aux))?|'
      r'à\s+la\s+charge\s+(du|de|des)|'
      r'de\s+la\s+faute\s+(du|de|des)|'
      r'engage\s+sa\s+responsabilité)(?![\p{L}\p{N}_])',
      caseSensitive: false,
      unicode: true,
    ),
  ),
  ForbiddenPattern(
    'INDEMNIFICATION_PROMISE',
    RegExp(
      // R0.2 (hotfix decouvert par execution reelle, troisieme vague) :
      // meme bug racine que les lookaround ci-dessus, mais a l'INTERIEUR
      // du motif cette fois - `\w` (utilise par `indemnis\w*`/
      // `dédommag\w*`) est LUI AUSSI ASCII-only en Dart (`[A-Za-z0-9_]`),
      // `unicode: true` ne change pas non plus ce comportement. Sur
      // "indemnisé" (avec un "é" APRES le radical "indemnis"), `\w*`
      // s'arretait avant le "é", puis le lookahead Unicode-aware
      // `(?![\p{L}\p{N}_])` juste apres refusait la position car "é" EST
      // un caractere de mot Unicode (`\p{L}`) - contradiction entre un
      // `\w*` ASCII et une frontiere verifiee en Unicode. Preuve reelle :
      // le test "sera indemnisé intégralement" ne levait plus
      // l'exception (`Actual: returned <null>`). Corrige en remplacant
      // `\w*` par `[\p{L}\p{N}_]*`, Unicode-aware au meme titre que les
      // lookaround de frontiere.
      r'(?<![\p{L}\p{N}_])(indemnis[\p{L}\p{N}_]*|dédommag[\p{L}\p{N}_]*|remboursera|'
      r'remboursement\s+(dû|du|garanti)|'
      r'prise\s+en\s+charge\s+financière)(?![\p{L}\p{N}_])',
      caseSensitive: false,
      unicode: true,
    ),
  ),
  ForbiddenPattern(
    'LEGAL_CONCLUSION',
    RegExp(
      r'(?<![\p{L}\p{N}_])(manquement\s+contractuel|inexécution\s+contractuelle|'
      r'violation\s+du\s+contrat|vice\s+caché|négligence|'
      r"défaut\s+d.exécution|obligation\s+de\s+résultat)(?![\p{L}\p{N}_])",
      caseSensitive: false,
      unicode: true,
    ),
  ),
  ForbiddenPattern(
    'INVENTED_AMOUNT',
    RegExp(
      r'(\d+[.,]?\d*\s?(€|eur\b|euros?)|montant\s+de\s+\d)',
      caseSensitive: false,
      unicode: true,
    ),
  ),
  ForbiddenPattern(
    'LEGAL_QUALIFICATION',
    RegExp(
      r'(?<![\p{L}\p{N}_])(délit|infraction|faute\s+lourde|force\s+majeure|'
      r'non[\s-]conformité\s+contractuelle|vice\s+de\s+forme)(?![\p{L}\p{N}_])',
      caseSensitive: false,
      unicode: true,
    ),
  ),
];

String? _fieldValue(ConfirmedFactData fact, String fieldName) {
  switch (fieldName) {
    case 'packagingCondition':
      return fact.packagingCondition;
    case 'productCondition':
      return fact.productCondition;
    case 'locationOnItem':
      return fact.locationOnItem;
    case 'productLabel':
      return fact.productLabel;
    case 'productReference':
      return fact.productReference;
    default:
      return null;
  }
}

/// Lève [LiabilityAttributionException] si un champ texte libre de [fact]
/// contient un motif interdit. Ne modifie jamais la donnée.
void screenConfirmedFact(ConfirmedFactData fact) {
  for (final String fieldName in liabilityGuardFreeTextFields) {
    final String? value = _fieldValue(fact, fieldName);
    if (value == null || value.isEmpty) {
      continue;
    }
    for (final ForbiddenPattern forbidden in forbiddenPatterns) {
      final RegExpMatch? match = forbidden.pattern.firstMatch(value);
      if (match != null) {
        throw LiabilityAttributionException(
          "Le champ '$fieldName' contient un contenu interdit dans une "
          'réserve (${forbidden.violationCode}) : « ${match.group(0)} ». '
          'Reformulez ce champ en constat factuel neutre (état physique '
          'observé uniquement), sans attribution de responsabilité, '
          "promesse d'indemnisation, conclusion ou qualification juridique, "
          'ni montant.',
          violationCode: forbidden.violationCode,
          fieldName: fieldName,
        );
      }
    }
  }
}

/// Applique [screenConfirmedFact] à chaque fait. Utilisé en défense en
/// profondeur juste avant `reserve_composer.composeReserve()`.
void screenConfirmedFacts(List<ConfirmedFactData> facts) {
  for (final ConfirmedFactData fact in facts) {
    screenConfirmedFact(fact);
  }
}

File: backend/app/domain/liability_guard.py
"""Garde-fou deterministe anti-attribution de responsabilite (R0.1, point 8
de la demande corrective "Local-First").

Bug corrige ici : la baseline R0 laissait `packaging_condition` et
`product_condition` (champs texte libre de `ConfirmedFactData`) totalement
libres de contenu. Rien n'empechait un utilisateur - ou une reparation IA
mal cadree en amont - de saisir par exemple :

    packaging_condition = "transporteur responsable"

Ce texte etant confirme (`user_confirmed=True`), il passait tel quel dans
`app.domain.templates.fr_v1._condition_clause()` et ressortait donc,
verbatim, dans la reserve finale generee par
`app.domain.reserve_composer.compose_reserve()` - alors meme que la GATE
zero-invention (section 2.4) et l'interdiction stricte du cahier des charges
("Aucune reserve ne doit jamais contenir : une attribution de faute ou de
responsabilite, une promesse d'indemnisation, une conclusion juridique, un
montant de reparation ou d'indemnisation invente, une qualification
juridique creee par l'IA") s'appliquent A LA RESERVE FINALE, quelle que soit
la provenance du texte (utilisateur, IA, ou les deux).

Principe de la correction : un ConfirmedFactData "confirme par l'utilisateur"
(invariant 6.3) garantit uniquement que l'utilisateur a valide la VALEUR du
champ, pas que cette valeur respecte le perimetre autorise d'une reserve
descriptive (constat factuel neutre). Ce module ajoute donc une deuxieme
barriere, independante de la confirmation utilisateur, appliquee :

  1. au moment de la confirmation (F09, voir
     app.application.use_cases.confirm_facts) pour un retour immediat a
     l'utilisateur ;
  2. juste AVANT app.domain.reserve_composer.compose_reserve(), en defense
     en profondeur - explicitement demande par le cahier des charges
     ("garde-fou deterministe AVANT le Reserve Composer") - de sorte
     qu'aucun chemin de code, present ou futur, ne puisse contourner l'appel
     (1) et laisser passer un contenu interdit.

Deterministe : regex figees, aucun appel IA/reseau. Memes entrees => meme
verdict a l'identique, pour toute relecture/audit (meme invariant que le
Reserve Composer lui-meme, cf. app/domain/reserve_composer.py).

Volontairement PAS de sanitization automatique (on ne reecrit jamais
silencieusement le texte d'un utilisateur) : le systeme rejette et demande a
l'utilisateur de reformuler en constat factuel neutre, conformement au
principe "aucune conclusion juridique generee par le systeme" (section 1.4).
"""

from __future__ import annotations

import re
from collections.abc import Sequence

from app.domain.errors import LiabilityAttributionError
from app.domain.fact_set import ConfirmedFactData

# Champs texte libre de ConfirmedFactData susceptibles de contenir une
# formulation utilisateur non contrainte (les autres champs sont soit des
# enums/nombres, soit calcules par code - voir app/domain/fact_set.py).
FREE_TEXT_FIELDS: tuple[str, ...] = (
    "packaging_condition",
    "product_condition",
    "location_on_item",
    "product_label",
    "product_reference",
)

# Chaque motif est deliberement large (faux positifs > faux negatifs) : ces
# champs ne doivent decrire QUE un etat physique constate ("carton enfonce",
# "produit raye"), jamais une cause, une consequence juridique/financiere ou
# une partie responsable. Un rejet legitime mais trop strict se corrige en
# reformulant le constat ; un contenu interdit qui fuite dans une reserve
# signee ne se corrige pas.
#
# R2 (voir app/domain/candidate_guard.py) : ces motifs sont desormais
# PUBLICS et partages avec le garde-fou applique aux CandidateFactData
# (sortie brute IA, avant toute confirmation utilisateur). Une seule source
# de verite pour "qu'est-ce qu'un contenu interdit dans une reserve" evite
# qu'un motif ajoute/corrige d'un cote (ex: nouvelle formulation de
# responsabilite reperee lors du benchmark R2) ne soit pas applique de
# l'autre.
FORBIDDEN_PATTERNS: dict[str, re.Pattern[str]] = {
    "LIABILITY_ATTRIBUTION": re.compile(
        r"\b(responsab(?:le|ilité)s?|fautifs?|\bfaute\b|"
        r"imputable(?:\s+(?:au|à|aux))?|"
        r"à\s+la\s+charge\s+(?:du|de|des)|"
        r"de\s+la\s+faute\s+(?:du|de|des)|"
        r"engage\s+sa\s+responsabilité)\b",
        re.IGNORECASE,
    ),
    "INDEMNIFICATION_PROMISE": re.compile(
        r"\b(indemnis\w*|dédommag\w*|remboursera|"
        r"remboursement\s+(?:dû|du|garanti)|"
        r"prise\s+en\s+charge\s+financière)\b",
        re.IGNORECASE,
    ),
    "LEGAL_CONCLUSION": re.compile(
        r"\b(manquement\s+contractuel|inexécution\s+contractuelle|"
        r"violation\s+du\s+contrat|vice\s+caché|négligence|"
        r"défaut\s+d.exécution|obligation\s+de\s+résultat)\b",
        re.IGNORECASE,
    ),
    "INVENTED_AMOUNT": re.compile(
        r"(\d+[.,]?\d*\s?(?:€|eur\b|euros?)|montant\s+de\s+\d)",
        re.IGNORECASE,
    ),
    "LEGAL_QUALIFICATION": re.compile(
        r"\b(délit|infraction|faute\s+lourde|force\s+majeure|"
        r"non[\s-]conformité\s+contractuelle|vice\s+de\s+forme)\b",
        re.IGNORECASE,
    ),
}


def screen_confirmed_fact(fact: ConfirmedFactData) -> None:
    """Leve LiabilityAttributionError si un champ texte libre de `fact`
    contient un motif interdit. Ne modifie jamais la donnee (pas de
    sanitization silencieuse)."""
    for field_name in FREE_TEXT_FIELDS:
        value = getattr(fact, field_name, None)
        if not value or not isinstance(value, str):
            continue
        for violation_code, pattern in FORBIDDEN_PATTERNS.items():
            match = pattern.search(value)
            if match:
                raise LiabilityAttributionError(
                    f"Le champ '{field_name}' contient un contenu interdit dans "
                    f"une reserve ({violation_code}) : « {match.group(0)} ». "
                    "Reformulez ce champ en constat factuel neutre (etat physique "
                    "observe uniquement), sans attribution de responsabilite, "
                    "promesse d'indemnisation, conclusion ou qualification "
                    "juridique, ni montant.",
                    violation_code=violation_code,
                    field_name=field_name,
                )


def screen_confirmed_facts(facts: Sequence[ConfirmedFactData]) -> None:
    """Applique `screen_confirmed_fact` a chaque fait. Utilise en defense en
    profondeur juste avant `reserve_composer.compose_reserve()` (voir
    docstring du module)."""
    for fact in facts:
        screen_confirmed_fact(fact)
